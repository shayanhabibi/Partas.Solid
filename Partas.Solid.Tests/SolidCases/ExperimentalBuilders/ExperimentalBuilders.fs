module Partas.Solid.Tests.SolidCases.ExperimentalBuilders

open Partas.Solid
open Partas.Solid.Experimental
open Fable.Core
open Fable.Core.JS

[<JS.Pojo>]
type MyObject(value: int, label: string) =
    member val value = value with get,set
    member val label = label with get,set

let getter,setter = createSignal(10)
let data,store = createStore<MyObject ResizeArray>(ResizeArray([||]))

effect {
    if getter() = 10 then
        store.Update(ResizeArray([||]))
    else
        setter 10
}
let getGetter = lambda {
    let mutable check = 10
    match getter() with
    | 10 -> check <- 5
    | 5 -> check <- 10
    check
}

let getSetter = lambda {
    store.Update(fun o -> o.Add(MyObject(5, "Test")); o)
    if data.Count > 0 then
        setter
}

[<SolidComponent>]
let ComponentWrap () =
    effect {
        if getter() = 10 then
            store.Update(ResizeArray([||]))
        else
            setter 10
    }
    let getGetter = lambda {
        let mutable check = 10
        match getter() with
        | 10 -> check <- 5
        | 5 -> check <- 10
        check
    }

    let getSetter = lambda {
        store.Update(fun o -> o.Add(MyObject(5, "Test")); o)
        if data.Count > 0 then
            setter
    }
    
    if getter() = 0 then getGetter
    else unbox getSetter
    
[<Erase>]
type TestComponent() =
    interface RegularNode
    [<SolidTypeComponent>]
    member props.__ =
        let childs = children { props.children }
        effect {
            printfn "effect"
        }
        cleanup {
            printfn "cleanup"
        }
        mount {
            printfn "mount"
        }
        let x = memo {
            printfn "memo"
            "memo"
        }
        let submission = batch {
            printfn ""
            childs()
        }
        let isSelected = selector {
            false
        }
        
        div() { childs() }
    
