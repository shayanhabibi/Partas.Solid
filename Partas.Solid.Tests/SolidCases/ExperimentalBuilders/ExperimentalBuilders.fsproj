<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net9.0</TargetFramework>
    </PropertyGroup>

    <ItemGroup>
        <Compile Include="ExperimentalBuilders.fs" />
    </ItemGroup>

    <ItemGroup>
        <ProjectReference Include="..\..\..\Partas.Solid\Partas.Solid.fsproj" />
    </ItemGroup>

</Project>
