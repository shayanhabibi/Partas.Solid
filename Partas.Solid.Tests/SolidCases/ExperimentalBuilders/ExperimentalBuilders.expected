import { createSelector, batch, createMemo, onMount, onCleanup, children, splitProps, createEffect, createSignal } from "solid-js";
import { createStore } from "solid-js/store";
import { printf, toConsole } from "./fable_modules/fable-library-js.5.0.0-alpha.13/String.js";

export const patternInput$004013 = createSignal(10);

export const setter = patternInput$004013[1];

export const getter = patternInput$004013[0];

export const patternInput$004014$002D1 = createStore([]);

export const store = patternInput$004014$002D1[1];

export const data = patternInput$004014$002D1[0];

createEffect(() => {
    (getter() === 10) ? ((store([]), (value) => {
    })) : ((setter(10), (value_1) => {
    }));
});

export const getGetter = () => {
    let matchValue;
    let check = 10;
    (matchValue = (getter() | 0), (matchValue === 5) ? ((check = 10, (value_1) => {
    })) : ((matchValue === 10) ? ((check = 5, (value) => {
    })) : ((value_2) => {
    })));
    return check | 0;
};

export const getSetter = () => {
    store((o) => {
        void (o.push({
            value: 5,
            label: "Test",
        }));
        return o;
    });
    return (data.length > 0) ? setter : ((value) => {
    });
};

export function ComponentWrap() {
    createEffect(() => {
        if (getter() === 10) {
            store([]);
        }
        else {
            setter(10);
        }
    });
    if (getter() === 0) {
        return () => {
            let check = 10;
            const matchValue = getter() | 0;
            switch (matchValue) {
                case 5: {
                    check = 10;
                    break;
                }
                case 10: {
                    check = 5;
                    break;
                }
                default: {
                }
            }
            return check | 0;
        };
    }
    else {
        return () => ((() => {
            store((o) => {
                void (o.push({
                    value: 5,
                    label: "Test",
                }));
                return o;
            });
            if (data.length > 0) {
                return setter;
            }
            else {
                return undefined;
            }
        })() | 0);
    }
}

export function TestComponent(props) {
    const [PARTAS_LOCAL, PARTAS_OTHERS] = splitProps(props, ["children"]);
    const childs = children(() => PARTAS_LOCAL.children);
    createEffect(() => {
        toConsole(printf("effect"));
    });
    onCleanup(() => {
        toConsole(printf("cleanup"));
    });
    onMount(() => {
        toConsole(printf("mount"));
    });
    const x = createMemo(() => {
        toConsole(printf("memo"));
        return "memo";
    });
    const submission = batch(() => {
        toConsole(printf(""));
        return childs();
    });
    const isSelected = createSelector(() => false);
    return <div>
        {childs()}
    </div>;
}

